File: src/FSharpPolymorphism/Animal.fs
namespace FSharpPolymorphism

type Animal =
    | Dog
    | Cat
    | Duck

module Animal =
    let makeNoise animal =
        match animal with
        | Dog -> "Bark"
        | Cat -> "Meow"
        | Duck -> "Quack"

    let fly animal =
        match animal with
        | Duck -> Some "The duck flies away!"
        | Dog
        | Cat -> None

File: src/CompositionBasedPolymorphism/Animal.cs
namespace CompositionBasedPolymorphism
{
    public interface IAnimal
    {
    }

    public interface IMakesNoise
    {
        string MakeNoise();
    }

    public interface ICanFly
    {
        string Fly();
    }

    public class Dog : IAnimal, IMakesNoise
    {
        public string MakeNoise() => "Bark";
    }

    public class Cat : IAnimal, IMakesNoise
    {
        public string MakeNoise() => "Meow";
    }

    public class Duck : IAnimal, IMakesNoise, ICanFly
    {
        public string MakeNoise() => "Quack";

        public string Fly() => "The duck flies away!";
    }
}
